Pp/rlhf

pascal-pfeiffer opened the pull request:
This PR adds RLHF as an option for fine-tuning a LLM.
It makes heavy use of the awesome [trl](https://github.com/lvwerra/trl) lib.

- [x] RLHF on single GPU setups
- [x] shared weights between reference model and active model to reduce the huge memory requirement when having three models running
- [x] DDP support
- [ ] Train your own reward model


Some logs to show that it is possible to overfit the reward model (though it doesn't translate to any increase in GPT score, yet):

Reward:
![train_ppo_mean_scores](https://github.com/h2oai/h2o-llmstudio/assets/1069138/69b9d8c3-d346-4bb0-8e7b-6f80479c7c1f)

GPT3.5 score:
![validation_GPT3 5](https://github.com/h2oai/h2o-llmstudio/assets/1069138/40e54402-31a5-4886-b91e-6ec99381f39f)

Validation loss:
![validation_loss](https://github.com/h2oai/h2o-llmstudio/assets/1069138/6a3b7134-fae3-4d5a-a85e-d96a6754e615)

[internal logs] https://app.neptune.ai/o/Zoo/org/h2o-llm/runs/details?viewId=99593e71-d48c-4a30-827b-b5efe805f0de&detailsTab=charts&shortId=HOL-1534

maxjeblick reviewed (commented):
Thanks a lot for this awesome new feature!

I'm currently still going trough the code (looks good so far!), some preliminary feedback:

- I'm getting WARNING: `use_cache=True` is incompatible with gradient check pointing. Setting `use_cache=False` warnings throughout the training (default settings with `h2oai/h2ogpt-gm-oasst1-en-2048-falcon-7b-v2`).
- We probably should fail ddp training as it is not yet supported.
- There are some parameters like gradient accumulation that are not used for rlhf. Would it be possible to hide those when rlhf is enabled (never checked if it's possible with nesting)?

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -172,15 +177,17 @@ def get_validation_collate_fn(self):
         output["predicted_text"] = np.array(predicted_text)
-        del output["predicted_answer_ids"]
+
+        if not cfg.training.use_rlhf:
```
Do you think it makes sense to repopulate `predicted_answer_ids` in `train.py` instead?

pascal-pfeiffer replied:
You mean making a copy of the `predicted_answer_ids` first and then add back to the dict after the post processing? sounds a bit complicated to me, or I am probably missing something.

maxjeblick replied:
Should be fine the way it is currently

maxjeblick commented on train.py:
```diff
@@ -387,7 +480,7 @@ def run_train(
     if cfg.environment._distributed:
         torch.distributed.barrier()
 
-    return val_data, val_loss, val_metric, batch
```
Nice refactoring! In line 203, `batch = None` can be deleted, as well.

maxjeblick commented on train.py:
```diff
@@ -500,12 +594,24 @@ def run(cfg: Any) -> None:
 
     model.to(cfg.environment._device)
+    if cfg.training.use_rlhf:
+        if cfg.training.offload_reward_model:
```
Like the offloading!

maxjeblick commented on llm_studio/src/trl/trainer.py:
```diff
@@ -0,0 +1,851 @@
+            all_logprobs, _, values, masks = self.batched_forward_pass(
+                self.model, queries, responses, model_inputs
+            )
+            with self.model.backbone.disable_adapter():
```
Awesome method to avoid having the reference model as a separate model.
We probably should fail if LORA is disabled or force enable lora with a warning in the logs.

maxjeblick replied:
Also, force embedding gradients should be disabled.

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -322,12 +339,18 @@ def _read_data(self, idx: int, sample: Dict) -> Dict:
-            ]
-        ).to(torch.bool)
+
+        if self.cfg.training.use_rlhf and self.mode == "train":
```
The `_read_data` method is a bit hard to follow. But I can also refactor after this PR has been merged.

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -294,12 +301,17 @@ def _get_sample(self, idx):
+        if self.cfg.training.use_rlhf and self.mode == "train":
+            return [prompt_encodings]
+        else:
+            return [prompt_encodings, answer_encodings]
```
Answer encodings would still be needed for the chat history (when using `parent_id_column`)?

pascal-pfeiffer replied:
In this case, afaik, the "old" answers are in the prompt. But it's definetly tricky, as they need to be passed to the reward model, too. Which is not really trained on full conversations but rather single question, single answer if I see that correctly.

maxjeblick commented on llm_studio/src/trl/trainer.py:
```diff
@@ -0,0 +1,851 @@
+    return logpy
+
+
+def whiten(values, shift_mean=True):
```
Seems to be unused

maxjeblick commented on llm_studio/src/models/text_causal_language_modeling_model.py:
```diff
@@ -142,66 +220,259 @@ def generate(self, batch: Dict, cfg: Any):
         return outputs
+
+
+class GPTNeoXRewardModelConfig(GPTNeoXConfig):
```
Could we move the reward models to a separate file?

maxjeblick commented on llm_studio/python_configs/text_causal_language_modeling_config.py:
```diff
@@ -196,6 +234,24 @@ def __post_init__(self):
+                "ppo_clip_value",
+                "scaling_factor_value_loss",
+                "ppo_epochs",
+                "ppo_batch_size",
```
`ppo_generate_temperature` is missing here

maxjeblick commented on llm_studio/src/models/text_causal_language_modeling_model.py:
```diff
@@ -142,66 +220,259 @@ def generate(self, batch: Dict, cfg: Any):
-        calculate_loss: bool = True,
+        generate: bool = False,
+        padding: bool = True,
+        is_ref_model: bool = False,
```
I think one can remove `is_ref_model` flag and also compute loss/value in the reference model (they will not be used afterwards, of course). This would result in slightly slimmer code. WDYT?

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -322,12 +339,18 @@ def _read_data(self, idx: int, sample: Dict) -> Dict:
+            ).to(torch.bool)
         attention_mask = torch.ones_like(input_ids)
 
         labels = input_ids.clone()
```
We could think about skipping label creation for rlhf (I should not be needed for rlhf).

maxjeblick commented on train.py:
```diff
@@ -259,59 +272,139 @@ def run_train(
+                logger.debug("Rollout: Generating response from active model")
+                output_dict = {}
+                output_dict["predicted_answer_ids"] = model.generate(
+                    batch, model.cfg, remove_prompt=True
```
I noticed that I mixed up `max_length_answer` with min length and max length for inference (causing OOMs).

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -62,7 +62,9 @@ def __init__(self, df: pd.DataFrame, cfg: Any, mode: str = "train"):
 
-        self.prompts = get_texts(df, self.cfg, separator="")
+        self.raw_prompts = get_texts(df, self.cfg, separator="")
+        self.prompts = [self.parse_prompt(cfg, prompt) for prompt in self.raw_prompts]
```
It is possible to move `.parse_prompt` to the `_get_sample` method and to the example logging.
By this, one does not need to keep two versions of the prompts.

pascal-pfeiffer replied:
we use `parse_prompt` in the chat

maxjeblick replied:
Ah sorry, I mistyped my idea:
What I meant was to remove `self.prompts` (and rename `self.raw_prompts` to `self.prompts`) and call `self.parse_prompt` in the `_get_sample` method and when logging an example.

maxjeblick commented on llm_studio/python_configs/text_causal_language_modeling_config.py:
```diff
@@ -158,16 +174,24 @@ def __post_init__(self):
+                "OpenAssistant/oasst-rm-2.1-pythia-1.4b-epoch-2.5",
+                "OpenAssistant/oasst-rm-2-pythia-6.9b-epoch-1",
+            ),
+            allow_custom=True,
```
`allow_custom=False` as `get_score` method currently only supports the models listed above.

maxjeblick commented on llm_studio/src/models/text_causal_language_modeling_model.py:
```diff
@@ -146,67 +214,114 @@ def generate(self, batch: Dict, cfg: Any):
+            output = output[:, input_ids.shape[1] :]
+        else:
+            # Mask the prompt tokens
+            output[:, : input_ids.shape[1]] = pad_token_id
```
Would it be possible to always remove the prompt?
We still mask it, and decoding in  `postprocess_batch_predictions` will skip pad tokens.

pascal-pfeiffer replied:
yes, probably makes more sense. i didn't want to change the original behavior, but this looks reasonable.

maxjeblick commented on llm_studio/src/datasets/text_causal_language_modeling_ds.py:
```diff
@@ -345,6 +358,11 @@ def _read_data(self, idx: int, sample: Dict) -> Dict:
                 ):
                     break
                 samples.insert(0, self._get_sample(int(parent_idx)))
+                sample["reward_model_prompt_text"] = (
```
As discussed, it should probably look something like
```
sample["reward_model_prompt_text"] = (
    self.raw_prompts[int(parent_idx)] + "<|endoftext|>" + self.answers[int(parent_idx)]
    + "<|endoftext|>"  # this is replaced later in the pipeline
    + sample["reward_model_prompt_text"])
```

maxjeblick commented:
Another thing I noticed is the scheduler looks a bit weird. It should be `cos(x)` with `x \in [0, pi/4]` for the whole training?
![Screenshot 2023-06-20 at 19-28-17 H2O LLM Studio](https://github.com/h2oai/h2o-llmstudio/assets/24281881/aec630d2-d05e-4d79-80fd-eb45b69ea9e9)

pascal-pfeiffer commented:
Thanks a ton for your great review and all the testing. 
Great catches!

I addressed your comments and followed most of your suggestions. 
I would also like to refactor the backward pass into a reusable method, but I would do it later in subsequent issues.

maxjeblick reviewed (approved):
LGTM, thanks again for the great work!
